solr/core/src/java/org/apache/solr/search/VectorSimilaritySourceParser.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.search;
+
+import static org.apache.solr.common.SolrException.ErrorCode.BAD_REQUEST;
+
+import java.util.Locale;
+import org.apache.lucene.index.VectorEncoding;
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.queries.function.ValueSource;
+import org.apache.lucene.queries.function.valuesource.ByteVectorSimilarityFunction;
+import org.apache.lucene.queries.function.valuesource.FloatVectorSimilarityFunction;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.schema.DenseVectorField;
+import org.apache.solr.schema.FieldType;
+import org.apache.solr.schema.SchemaField;
+
+/**
+ * This class provides implementation for two variants for parsing function query vectorSimilarity
+ * which is used to calculate the similarity between two vectors.
+ */
+public class VectorSimilaritySourceParser extends ValueSourceParser {
+  @Override
+  public ValueSource parse(FunctionQParser fp) throws SyntaxError {
+    final String arg1Str = fp.parseArg();
+    if (arg1Str == null || !fp.hasMoreArguments())
+      throw new SolrException(
+          BAD_REQUEST, "Invalid number of arguments. Please provide either two or four arguments.");
+    final boolean constVec = '[' == fp.sp.peek();
+    final String arg2Str = constVec ? null : fp.parseArg();
+    if (fp.hasMoreArguments() && arg2Str != null) return handle4ArgsVariant(fp, arg1Str, arg2Str);
+    return handle2ArgsVariant(fp, arg1Str, arg2Str, constVec);
+  }
+
+  private ValueSource handle4ArgsVariant(FunctionQParser fp, String arg1Str, String arg2Str)
+      throws SyntaxError {
+    final var vectorEncoding = VectorEncoding.valueOf(arg1Str);
+    final var vectorSimilarityFunction = VectorSimilarityFunction.valueOf(arg2Str);
```

Comment:
On a side note, VectorSimilarityFunction Enums may become obsolete in the future as they already made codec format incharge of it by moving the list of similarity into FieldsInfoFormat.
https://github.com/apache/lucene/pull/13119